File: n_queens.c
#include <stdio.h>
#include <stdlib.h>
#include <math.h>
#include <time.h>

int count = 0;

int is_safe(int *board, int row, int col) {
    for (int i = 0; i < row; i++) {
        if (board[i] == col || abs(board[i] - col) == abs(i - row))
            return 0;
    }
    return 1;
}

void solve(int *board, int row, int n) {
    if (row == n) {
        count++;
        return;
    }
    for (int col = 0; col < n; col++) {
        if (is_safe(board, row, col)) {
            board[row] = col;
            solve(board, row + 1, n);
            board[row] = -1;
        }
    }
}

int solve_nqueens(int n) {
    count = 0;
    int *board = malloc(sizeof(int) * n);
    for (int i = 0; i < n; i++) board[i] = -1;
    solve(board, 0, n);
    free(board);
    return count;
}

double now_ms() {
    return (double)clock() * 1000.0 / CLOCKS_PER_SEC;
}

int main() {
    int sizes[] = {4, 6, 8, 10, 12, 14};
    int reps = 15;
    int num_sizes = 6;
    
    printf("============================================================\n");
    printf("  PROBLEMA DAS N RAINHAS - BENCHMARK (C)\n");
    printf("============================================================\n");
    printf("\n");
    
    double all_means[6], all_stds[6];
    
    for (int s = 0; s < num_sizes; s++) {
        int n = sizes[s];
        printf("  Testando tabuleiro %dx%d... ", n, n);
        fflush(stdout);
        
        double *times = malloc(sizeof(double) * reps);
        for (int r = 0; r < reps; r++) {
            double t0 = now_ms();
            solve_nqueens(n);
            double t1 = now_ms();
            times[r] = t1 - t0;
        }
        
        double sum = 0, sum2 = 0;
        for (int i = 0; i < reps; i++) {
            sum += times[i];
            sum2 += times[i]*times[i];
        }
        double mean = sum / reps;
        double std = sqrt(sum2 / reps - mean*mean);
        free(times);
        
        all_means[s] = mean;
        all_stds[s] = std;
        
        printf("OK\n");
        printf("    Tempo medio: %10.2f ms\n", mean);
        printf("    Desvio padrao: %8.2f ms\n", std);
        printf("\n");
    }
    
    printf("============================================================\n");
    printf("  Resumo dos resultados:\n");
    printf("------------------------------------------------------------\n");
    printf("  %-10s %-20s %-20s\n", "Tamanho", "Tempo Medio (ms)", "Desvio Padrao (ms)");
    printf("------------------------------------------------------------\n");
    for (int s = 0; s < num_sizes; s++) {
        printf("  %dx%-6d %15.2f %18.2f\n", sizes[s], sizes[s], all_means[s], all_stds[s]);
    }
    printf("============================================================\n");
    printf("\n");
    
    return 0;
}

File: n_queens.py
import time
import statistics
import json

def solve_nqueens(n):
    board = [-1] * n
    solutions = []
    def is_safe(row, col):
        for i in range(row):
            if board[i] == col or abs(board[i]-col) == abs(i-row):
                return False
        return True
    def backtrack(row=0):
        if row == n:
            solutions.append(board[:])
            return
        for col in range(n):
            if is_safe(row, col):
                board[row] = col
                backtrack(row+1)
                board[row] = -1
    backtrack()
    return len(solutions)

def benchmark(n, reps=15):  
    times = []
    for _ in range(reps):
        t0 = time.perf_counter()
        solve_nqueens(n)
        t1 = time.perf_counter()
        times.append((t1 - t0) * 1000)
    return {"n": n, "mean_ms": statistics.mean(times), "std_ms": statistics.pstdev(times)}

if __name__ == "__main__":
    print("=" * 60)
    print("  PROBLEMA DAS N RAINHAS - BENCHMARK (Python)")
    print("=" * 60)
    print()
    
    results = []
    for n in [4, 6, 8, 10, 12, 14]:
        print(f"  Testando tabuleiro {n}x{n}...", end=" ", flush=True)
        res = benchmark(n)
        results.append(res)
        print(f"✓")
        print(f"    Tempo médio: {res['mean_ms']:>10.2f} ms")
        print(f"    Desvio padrão: {res['std_ms']:>8.2f} ms")
        print()
    
    print("=" * 60)
    print("  Resumo dos resultados:")
    print("-" * 60)
    print(f"  {'Tamanho':<10} {'Tempo Médio (ms)':<20} {'Desvio Padrão (ms)':<20}")
    print("-" * 60)
    for res in results:
        print(f"  {res['n']}x{res['n']:<6} {res['mean_ms']:>15.2f} {res['std_ms']:>18.2f}")
    print("=" * 60)
    print()
    print("  Resultados salvos em: tempos_python.json")
    print()
    
    with open("tempos_python.json", "w") as f:
        json.dump(results, f, indent=2)
